apps/explorer/src/components/OwnedCoins/components/OwnedCoinsPanel.tsx

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { type CoinStruct } from '@mysten/sui.js';
+import { useState, useEffect, useRef } from 'react';
+
+import CoinItem from './CoinItem';
+
+import { useGetCoins } from '~/hooks/useGetCoins';
+import { useOnScreen } from '~/hooks/useOnScreen';
+import { LoadingSpinner } from '~/ui/LoadingSpinner';
+
+type CoinsPanelProps = {
+    coinType: string;
+    id: string;
+};
+
+function CoinsPanel({ coinType, id }: CoinsPanelProps): JSX.Element {
+    const [coinObjects, setCoinObjects] = useState<CoinStruct[]>([]);
+    const containerRef = useRef(null);
+    const { isIntersecting } = useOnScreen(containerRef);
+    const { data, isLoading, isFetching, fetchNextPage, hasNextPage } =
+        useGetCoins(coinType, id);
+
+    useEffect(() => {
+        if (data) {
+            let coins: CoinStruct[] = [];
+            data.pages.forEach((page) => {
+                coins = [...coins, ...page.data];
+            });
+            setCoinObjects(coins);
+        }
+    }, [data]);
+
+    useEffect(() => {
+        if (isIntersecting && hasNextPage && !isFetching) {
+            fetchNextPage();
+        }
+    }, [isIntersecting, hasNextPage, isFetching, fetchNextPage]);
+
+    return (
+        <div>
+            {coinObjects.map((obj, index) => {
+                if (index === coinObjects.length - 1) {
```

Comment:
Rather than do this, I'd actually just render a div with the `LoadingSpinner` in it at the end of the list if `hasNextPage` is true, and ref that. That would simplify this logic a little bit.